### PR TITLE
Development

### PR DESCRIPTION
**Pull request by mossbeard**
New Rest Engine Service
I need a nuget package

### CODE REVIEW

File: src/Attributes/ValidateModelStateAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+                {
+                    if (argument != null)
+                        modelState.Root.ValidationState = ModelValidationState.Valid;
+                }
+            }
+        }
+
+        private object ParseDataType(string value)
+        {
+            if (Boolean.TryParse(value, out var boolResult))
```
Comment (konne) [inline src/Attributes/ValidateModelStateAttribute.cs:62]: häääää

this should work if (Boolean.TryParse(value, out var boolResult)
                               return boolResult;

File: src/Attributes/ValidateModelStateAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+                return Convert.ChangeType(value, typeof(bool));
```
Comment (konne) [inline src/Attributes/ValidateModelStateAttribute.cs]: reject -> INVARIANT CULTURE ???!!!!???

File: src/Controllers/FileOperations.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+            try
+            {
+                logger.Trace($"Start upload file - Id: {fileId}");
+                var result = PostUploadFile(serFilename, data, serUnzip, fileId);
+                logger.Trace($"{nameof(UploadWithId)} - Response: {JsonConvert.SerializeObject(result)}");
+                return new ObjectResult(result);
+            }
+            catch (Exception ex)
+            {
+                var msg = $"The Method {nameof(UploadWithId)} failed.";
```
Comment (konne) [inline src/Controllers/FileOperations.cs]: don't repeat yourself, use Methods

File: src/Controllers/FileOperations.cs [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        [Produces("application/json")]
+        [SwaggerOperation("Upload", "Upload a file to the service.", Tags = new[] { "Upload File" })]
+        [SwaggerResponse(statusCode: 200, type: typeof(OperationResult), description: "Returns a new generated file id.")]
+        public virtual IActionResult Upload([FromBody][Required] Stream data, [FromHeader][Required] string serFilename, [FromHeader] bool serUnzip = false)
+        {
+            try
+            {
+                logger.Trace("Start upload file");
+                var result = PostUploadFile(serFilename, data, serUnzip);
+                logger.Trace($"{nameof(Upload)} - Response: {JsonConvert.SerializeObject(result)}");
```
Comment (konne) [inline src/Controllers/FileOperations.cs]: same here Method ReturnObjectAndTraceToNLog(...)

File: src/Controllers/TaskOperations.cs [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+        [Produces("application/json")]
+        [SwaggerOperation("CreateTaskWithId", "Create a reporting task.", Tags = new[] { "Create Task" })]
+        [SwaggerResponse(statusCode: 200, type: typeof(OperationResult), description: "Returns a new generated file id.")]
+        public virtual IActionResult CreateTaskWithId([FromBody][Required] string jsonJob, [FromRoute][Required] Guid taskId)
+        {
+            try
+            {
+                logger.Trace($"Start create task - Json: {jsonJob}");
+                var result = PostStartTask(jsonJob, taskId);
+                logger.Trace($"{nameof(CreateTaskWithId)} - Response: {JsonConvert.SerializeObject(result)}");
```
Comment (konne) [inline src/Controllers/TaskOperations.cs]: same again, use helper Method

File: src/Controllers/TaskOperations.cs [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+                return new ObjectResult(result);
+            }
+            catch (Exception ex)
+            {
+                var msg = $"The Method {nameof(CreateTaskWithId)} failed.";
```
Comment (konne) [inline src/Controllers/TaskOperations.cs]: same here

File: src/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,20 @@
+FROM microsoft/dotnet:2.1-aspnetcore-runtime AS base
+WORKDIR /app
+EXPOSE 80
+EXPOSE 443
+
+FROM microsoft/dotnet:2.1-sdk AS build
```
Comment (konne) [inline src/Dockerfile]: MISSING 2.2

File: src/Filters/OpenApiDocumentFilter.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        /// </summary>
+        public void Apply(OpenApiDocument openApiDoc, DocumentFilterContext context)
+        {
+            try
+            {
+                foreach (var server in Servers)
+                {
+                    openApiDoc.Servers.Add(server);
+                    var serverUri = new Uri(server.Url);
+                    var basePath = serverUri?.PathAndQuery?.Split('?')?.FirstOrDefault() ?? null;
```
Comment (konne) [inline src/Filters/OpenApiDocumentFilter.cs]: Uri.Query / QueryBuilder / Parser

File: src/Program.cs [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+        public static void Main(string[] args)
+        {
+            try
+            {
+                //Activate Nlog logger with configuration
+                var logger = NLogBuilder.ConfigureNLog("App.config").GetCurrentClassLogger();
+
+                //Build config for webserver
+                var config = new ConfigurationBuilder()
+                .SetBasePath(AppContext.BaseDirectory)
```
Comment (konne) [inline src/Program.cs]: Indent

File: src/SerEngineRest.xml [hunk 10]
```diff
@@ -0,0 +1,422 @@
+<?xml version="1.0"?>
```
Comment (konne) [inline src/SerEngineRest.xml]: if this is agenrated file -> gitignore !!!

File: src/Services/ReportingService.cs [hunk 11]
```diff
@@ -0,0 +1,332 @@
...
+            }
+            catch (Exception ex)
+            {
+                logger.Error(ex, $"The task {taskId} failed.");
+            }
+        }
+
+        private void Delete(Guid? id = null)
+        {
+            if (id.HasValue)
```
Comment (konne) [inline src/Services/ReportingService.cs:137]: if you have id -> pick id -> fill folders with Array with one element -> default code

File: src/Services/ReportingService.cs [hunk 12]
```diff
@@ -0,0 +1,332 @@
...
+        /// <param name="taskId">Id of the task</param>
+        /// <param name="taskStatus">Select a special status</param>
+        /// <returns></returns>
+        public List<JobResult> GetTasks(Guid? taskId = null, TaskStatusInfo? taskStatus = null)
+        {
+            var results = new List<JobResult>();
+
+            try
+            {
+                if (taskId.HasValue)
```
Comment (konne) [inline src/Services/ReportingService.cs:246]: same as before

File: src/Startup.cs [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+                            {
+                                Name = "Sense Excel Reporting",
+                                Url = new Uri("http://senseexcel.com"),
+                            },
+                            License = new OpenApiLicense()
+                            {
+                                Name = "MIT"
+                            }
+                        });
+                        options.DescribeAllEnumsAsStrings();
```
Comment (konne) [inline src/Startup.cs:114]: try fluent way

File: src/ser-engine-rest.csproj [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    <PackageReference Include="Hjson" Version="3.0.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.1" />
+    <PackageReference Include="NLog" Version="4.5.11" />
+    <PackageReference Include="NLog.Extensions.Logging" Version="1.4.0" />
+    <PackageReference Include="NLog.Web.AspNetCore" Version="4.8.1" />
+    <PackageReference Include="Microsoft.AspNetCore.App" />
+    <PackageReference Include="Microsoft.SourceLink.GitHub" Version="1.0.0-beta-63127-02" PrivateAssets="All" />
+    <PackageReference Include="Swashbuckle.AspNetCore" Version="5.0.0-beta" />
+    <PackageReference Include="Swashbuckle.AspNetCore.Annotations" Version="5.0.0-beta" />
+    <PackageReference Include="System.Drawing.Common" Version="4.5.1" />
```
Comment (konne) [inline src/ser-engine-rest.csproj]: if this is in the engine I think you don't need it here

File: src/ser-engine-rest.xml [hunk 15]
```diff
@@ -0,0 +1,290 @@
+<?xml version="1.0"?>
```
Comment (konne) [inline src/ser-engine-rest.xml]: same-> autogenerated remove from git

File: src/wwwroot/README.md [hunk 16]
```diff
@@ -0,0 +1,42 @@
+# Welcome to ASP.NET 5 Preview
```
Comment (konne) [inline src/wwwroot/README.md]: NOP

File: src/wwwroot/index.html [hunk 17]
```diff
@@ -0,0 +1 @@
+<meta http-equiv="refresh" content="0;URL='./swagger/'" />
```
Comment (konne) [inline src/wwwroot/index.html]: ?? possible with kestrel config / redirect of routes?

File: src/wwwroot/web.config [hunk 18]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (konne) [inline src/wwwroot/web.config]: please google if remove is possible

**Review by konne — CHANGES_REQUESTED** [review]
@mossbeard please rework

**Comment by mossbeard** [comment]
@konne 
I have adjusted all comments now.
The "options" variable in the startup.cs can only be called individually.
Comment "try fluent way"
Please check again

**Review by konne — APPROVED** [review]
@mossbeard please merge

**Pull request merged**